catalog/app/containers/Bucket/PackageDialog/S3FilePicker.tsx

```diff
@@ -0,0 +1,342 @@
+
+  const [locked, setLocked] = React.useState(false)
+
+  const cancel = React.useCallback(() => onClose('cancel'), [onClose])
+
+  const handleClose = React.useCallback(
+    (_e: {}, reason: MuiCloseReason) => {
+      if (!locked) onClose(reason)
+    },
+    [locked, onClose],
+  )
+
+  const crumbs = React.useMemo(() => getCrumbs({ bucket, path }), [bucket, path])
+
+  const getCrumbLinkProps = ({ to }: { to: string }) => ({
+    onClick: () => {
+      setPath(to)
+    },
+  })
+
+  React.useLayoutEffect(() => {
+    // reset accumulated results when bucket, path and / or prefix change
+    setPrev(null)
+  }, [bucket, path, prefix])
+
+  const data = useData(bucketListing, { bucket, path, prefix, prev })
+
+  const loadMore = React.useCallback(() => {
+    AsyncResult.case(
+      {
+        Ok: (res: requests.BucketListingResult) => {
+          // this triggers a re-render and fetching of next page of results
+          if (res.continuationToken) setPrev(res)
+        },
+        _: () => {},
+      },
+      data.result,
+    )
+  }, [data.result])
+
+  const add = React.useCallback(() => {
+    data.case({
+      Ok: async (r: requests.BucketListingResult) => {
+        try {
+          setLocked(true)
+          const dirsByBasename = R.fromPairs(
+            r.dirs.map((name) => [ensureNoSlash(withoutPrefix(r.path, name)), name]),
+          )
+          const filesByBasename = R.fromPairs(
+            r.files.map((f) => [withoutPrefix(r.path, f.key), f]),
+          )
+          const { dirs, files } = selection.reduce(
+            (acc, id) => {
+              const dir = dirsByBasename[id]
+              if (dir) return { ...acc, dirs: [...acc.dirs, dir] }
+              const file = filesByBasename[id]
+              if (file) return { ...acc, files: [...acc.files, file] }
+              return acc // shouldnt happen
+            },
+            { files: [] as requests.BucketListingFile[], dirs: [] as string[] },
+          )
+
+          // TODO: drain?
+          const dirsPromises = dirs.map((dir) =>
+            limit(bucketListing, { bucket, path: dir, delimiter: false }),
+          )
+
+          const dirsChildren = await Promise.all(dirsPromises)
+          const allChildren = dirsChildren.reduce(
+            (acc, res) => acc.concat(res.files),
+            [] as requests.BucketListingFile[],
+          )
+
+          onClose({ files: files.concat(allChildren), path })
+        } finally {
+          setLocked(false)
+        }
+      },
+      _: () => {},
+    })
+  }, [onClose, selection, data, bucket, path, bucketListing])
+
+  const handleExited = React.useCallback(() => {
+    setPath('')
+    setPrefix('')
+    setPrev(null)
+    setSelection([])
+  }, [])
+
+  return (
+    <M.Dialog
+      open={open}
+      onClose={handleClose}
+      onExited={handleExited}
+      fullWidth
+      maxWidth="lg"
+      // TODO: height?
+      // scroll="body"
```

Comment:
Yes, I think that table should be scrolled (ideally, the table's content), but buttons "Cancel"/"Add files" always visible

Reply:
yep, i'm trying to implement this behaviour